clang.go

```diff
@@ -0,0 +1,286 @@
+	repoPath, err := repos.GetRepoPathByName(codeFile.RepoName)
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+	absRepoPath, err := filepath.Abs(string(repoPath))
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+
+	pathInRepo, err := repos.PathInRepo(codeFile.RepoName, codeFile.Path)
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+
+	pathInRepo, err = filepath.Abs(pathInRepo)
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+
+	commands := db.AllCompileCommands()
+	defer commands.Dispose()
+	cmdline := []string{}
+	command := findMatchingCommand(pathInRepo, commands)
+	buildDir := absRepoPath
+	if command != nil {
+		for i := uint32(0); i < command.NumArgs(); i++ {
+			cmdline = append(cmdline, command.Arg(i))
+		}
+		buildDir = command.Directory()
+	}
+
+	// We need to check the translation unit in the build directory
+	originalDir, err := os.Getwd()
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+	err = os.Chdir(buildDir)
+	if err != nil {
+		return map[string]map[uint]*Code{}, err
+	}
+	defer os.Chdir(originalDir)
+
+	var tu clang.TranslationUnit
+	var clangErr clang.ErrorCode
+	if len(cmdline) != 0 {
+		clangErr = index.ParseTranslationUnit2FullArgv("", cmdline, nil, 0, &tu)
+	} else {
+		clangErr = index.ParseTranslationUnit2(pathInRepo, clangArgs, nil, 0, &tu)
+	}
+	if clangErr != clang.Error_Success {
+		return map[string]map[uint]*Code{}, fmt.Errorf("Error parsing translation unit `%s`, %v\n", codeFile.Path, clangErr)
+	}
+	defer tu.Dispose()
+
+	for _, d := range tu.Diagnostics() {
+		fmt.Printf("Diagnostic for file %s: %s\n", codeFile.Path, d.Spelling())
+	}
+
+	return visitAstNodes(tu.TranslationUnitCursor(), codeFile.RepoName, absRepoPath, codeFile.Path, fileMap), nil
+
+}
+
+// Tags the code in the given repository using libclang. The compilationDatabase path and clang arguments are optional
+// and used to provide libclang as much information as possible when parsing the code. This function will parse each file individually,
+// but collect tagged data from all included files. This helps to tag code from header files that normally is
+// not found in the compilation database (because it is only part of a translation unit as a result of being included from other files)
+// @llr REQ-TRAQ-SWL-61
+func tagCodeLibClang(repoName repos.RepoName, codeFiles []CodeFile, compilationDatabase string, clangArgs []string) (map[CodeFile][]*Code, error) {
+	codeMap := make(map[string]map[uint]*Code)
+	code := make(map[CodeFile][]*Code)
+
+	index := clang.NewIndex(1, 0)
+	defer index.Dispose()
+
+	var compDb clang.CompilationDatabase
+	if compilationDatabase != "" {
+		pathInRepo, err := repos.PathInRepo(repoName, compilationDatabase)
+		if err != nil {
+			fmt.Printf("Error opening compilation database in path `%s`: `%v`", compilationDatabase, err)
```

Comment:
I don't want to make this a hard error (although I have been going back and forth on this), because it just means that it will fall back to the clangArguments and could still potentially parse the code. I can rephrase this as a warning though, that would make it clearer.